internal/controllers/topology/cluster/reconcile_state.go

```diff
@@ -251,7 +251,8 @@ func (r *Reconciler) reconcileControlPlane(ctx context.Context, s *scope.Scope)
 	if s.Desired.ControlPlane.MachineHealthCheck != nil || s.Current.ControlPlane.MachineHealthCheck != nil {
 		// Set the ControlPlane Object and the Cluster as owners for the MachineHealthCheck to ensure object garbage collection
 		// in case something happens before the MHC sets ownership to the Cluster.
-		if s.Desired.ControlPlane.MachineHealthCheck != nil {
+		// NOTE: This is only necessary when the ControlPlane Object did not yet exist during computeDesiredState.
```

Comment:
With at least some backup/restore flows (e.g. velero) the owner references aren't restored - one reason for this is that uuids change - and are expected to be put back together by the controllers. 

I think I must have overlooked this in the original PR to add MHC to the ClusterClass, so it would be good to get it right now if we can.

Reply:
TLDR:

* wrong owner references would get garbage collected
* currently nothing re-adds the correct owner reference

Reply:
And same issue for MachineHealthChecks for MachineDeployments

Reply:
We could also keep the old behaviour by using:
```go
		if s.Desired.ControlPlane.MachineHealthCheck != nil {
			if s.Current.ControlPlane.Object == nil {
				// If the ControlPlane did not exist before have to use s.Desired.ControlPlane.Object
				s.Desired.ControlPlane.MachineHealthCheck.SetOwnerReferences([]metav1.OwnerReference{
					*ownerReferenceTo(s.Desired.ControlPlane.Object),
				})
			} else {
				// Otherwise we have to use s.Current.ControlPlane.Object because on no-ops s.Desired.ControlPlane.Object will 
				// not contain a uid.
				s.Desired.ControlPlane.MachineHealthCheck.SetOwnerReferences([]metav1.OwnerReference{
					*ownerReferenceTo(s.Current.ControlPlane.Object),
				})
			}
		}
```

Reply:
I would do the following:
* Drop mhc.SetOwnerReference from desired_state.go. This was only added to not accidentally drop the OwnerReference because we assumed we are not setting the OwnerReference somewhere else
* Set OwnerRef in reconcile_state.go. If possible from current otherwise from desired (like Christians suggestion)
    * This way we have a chance of getting the uit and avoiding another get call, otherwise we fall back to the get call
* Same for CP and MD MHCs

Reply:
Opened a PR here to show what I mean: https://github.com/kubernetes-sigs/cluster-api/pull/6660

Reply:
Talked to Fabrizio. The idea is now to keep the same logic but to implement it instead in desired_state.go to keep as much of the "desired state calculation" there

Reply:
#6660 should be ready now